File: backends/ktor/src/main/kotlin/com/hellotimepro/ktor/service/PlazaService.kt
package com.hellotimepro.ktor.service

import com.hellotimepro.ktor.db.dbQuery
import com.hellotimepro.ktor.domain.User
import com.hellotimepro.ktor.dto.CapsuleListItem
import com.hellotimepro.ktor.dto.Paginated
import com.hellotimepro.ktor.dto.Pagination
import com.hellotimepro.ktor.repository.CapsuleRepository
import com.hellotimepro.ktor.repository.FavoriteRepository
import com.hellotimepro.ktor.repository.PlazaSort
import com.hellotimepro.ktor.repository.UserRepository
import com.hellotimepro.ktor.web.ApiException
import java.util.UUID
import kotlin.math.ceil

class PlazaService(
    private val capsules: CapsuleRepository,
    private val users: UserRepository,
    private val favorites: FavoriteRepository,
    private val mapper: MapperService,
) {
    suspend fun plazaList(
        sort: String,
        filter: String,
        q: String?,
        page: Int,
        pageSize: Int,
        viewerId: UUID?,
    ): Paginated<CapsuleListItem> {
        Validation.page(page, pageSize)
        if (sort != "hot" && sort != "new") throw ApiException.validation("sort 仅支持 hot/new", "sort")
        if (filter != "all" && filter != "opened" && filter != "unopened") {
            throw ApiException.validation("filter 仅支持 all/opened/unopened", "filter")
        }
        var normalized = q?.trim()?.lowercase()
        if (normalized != null && normalized.isEmpty()) normalized = null
        if (normalized != null && normalized.length > 50) throw ApiException.validation("q 长度不得超过 50", "q")
        val qPattern = normalized?.let { "%$it%" }
        val sortEnum = if (sort == "hot") PlazaSort.HOT else PlazaSort.NEW
        val now = nowUtc()

        return dbQuery {
            val total = capsules.countPlaza(filter, now, qPattern)
            val rows = capsules.findPlazaPaged(filter, now, qPattern, sortEnum, offset(page, pageSize), pageSize)
            val owners = users.findAllByIds(rows.map { it.ownerId }).associateBy { it.id }
            val faved = loadFavoriteSet(viewerId, rows.map { it.id })
            val items = rows.mapNotNull { c ->
                owners[c.ownerId]?.let { mapper.listItem(c, it, c.id in faved, null) }
            }
            Paginated(items, pagination(total, page, pageSize))
        }
    }

    suspend fun myCapsules(user: User, page: Int, pageSize: Int): Paginated<CapsuleListItem> {
        Validation.page(page, pageSize)
        return dbQuery {
            val total = capsules.countByOwner(user.id)
            val rows = capsules.findByOwnerPaged(user.id, offset(page, pageSize), pageSize)
            val items = rows.map { mapper.listItem(it, user, favoritedByMe = false, favoritedAt = null) }
            Paginated(items, pagination(total, page, pageSize))
        }
    }

    suspend fun myFavorites(user: User, page: Int, pageSize: Int): Paginated<CapsuleListItem> {
        Validation.page(page, pageSize)
        return dbQuery {
            val total = favorites.countByUser(user.id)
            val favs = favorites.findByUserPaged(user.id, offset(page, pageSize), pageSize)
            if (favs.isEmpty()) {
                return@dbQuery Paginated(emptyList(), pagination(total, page, pageSize))
            }
            val capsuleMap = capsules.findAllByIds(favs.map { it.capsuleId }).associateBy { it.id }
            val owners = users.findAllByIds(capsuleMap.values.map { it.ownerId }).associateBy { it.id }
            val items = favs.mapNotNull { f ->
                val c = capsuleMap[f.capsuleId] ?: return@mapNotNull null
                owners[c.ownerId]?.let { mapper.listItem(c, it, favoritedByMe = true, favoritedAt = f.createdAt) }
            }
            Paginated(items, pagination(total, page, pageSize))
        }
    }

    private fun loadFavoriteSet(viewerId: UUID?, capsuleIds: List<UUID>): Set<UUID> {
        if (viewerId == null || capsuleIds.isEmpty()) return emptySet()
        return favorites.findByUserAndCapsuleIds(viewerId, capsuleIds).map { it.capsuleId }.toSet()
    }

    private fun offset(page: Int, pageSize: Int): Long = (page - 1).toLong() * pageSize

    private fun pagination(total: Long, page: Int, pageSize: Int): Pagination {
        val totalPages = if (pageSize == 0) 0 else ceil(total.toDouble() / pageSize).toInt()
        return Pagination(page, pageSize, total, totalPages)
    }
}

File: backends/aspnet/src/Services/PlazaService.cs
using HelloTimePro.Aspnet.Domain;
using HelloTimePro.Aspnet.Dto;
using HelloTimePro.Aspnet.Repositories;
using HelloTimePro.Aspnet.Web;

namespace HelloTimePro.Aspnet.Services;

public sealed class PlazaService
{
    private readonly CapsuleRepository _capsules;
    private readonly UserRepository _users;
    private readonly FavoriteRepository _favorites;
    private readonly MapperService _mapper;

    public PlazaService(
        CapsuleRepository capsules, UserRepository users, FavoriteRepository favorites, MapperService mapper)
    {
        _capsules = capsules;
        _users = users;
        _favorites = favorites;
        _mapper = mapper;
    }

    public async Task<Paginated<CapsuleListItem>> PlazaList(
        string sort, string filter, string? q, int page, int pageSize, Guid? viewerId)
    {
        Validation.Page(page, pageSize);
        if (sort != "hot" && sort != "new")
            throw ApiException.Validation("sort 仅支持 hot/new", "sort");
        if (filter != "all" && filter != "opened" && filter != "unopened")
            throw ApiException.Validation("filter 仅支持 all/opened/unopened", "filter");

        var normalized = q?.Trim().ToLowerInvariant();
        if (normalized is { Length: 0 }) normalized = null;
        if (normalized is { Length: > 50 })
            throw ApiException.Validation("q 长度不得超过 50", "q");

        var sortEnum = sort == "hot" ? PlazaSort.Hot : PlazaSort.New;
        var now = Times.NowUtc();

        var total = await _capsules.CountPlaza(filter, now, normalized);
        var rows = await _capsules.FindPlazaPaged(filter, now, normalized, sortEnum, Offset(page, pageSize), pageSize);
        var owners = (await _users.FindAllByIds(rows.Select(r => r.OwnerId).ToList())).ToDictionary(u => u.Id);
        var faved = await LoadFavoriteSet(viewerId, rows.Select(r => r.Id).ToList());

        var items = rows
            .Where(c => owners.ContainsKey(c.OwnerId))
            .Select(c => _mapper.ListItem(c, owners[c.OwnerId], faved.Contains(c.Id), null))
            .ToList();
        return new Paginated<CapsuleListItem>(items, Page(total, page, pageSize));
    }

    public async Task<Paginated<CapsuleListItem>> MyCapsules(User user, int page, int pageSize)
    {
        Validation.Page(page, pageSize);
        var total = await _capsules.CountByOwner(user.Id);
        var rows = await _capsules.FindByOwnerPaged(user.Id, Offset(page, pageSize), pageSize);
        var items = rows.Select(c => _mapper.ListItem(c, user, favoritedByMe: false, favoritedAt: null)).ToList();
        return new Paginated<CapsuleListItem>(items, Page(total, page, pageSize));
    }

    public async Task<Paginated<CapsuleListItem>> MyFavorites(User user, int page, int pageSize)
    {
        Validation.Page(page, pageSize);
        var total = await _favorites.CountByUser(user.Id);
        var favs = await _favorites.FindByUserPaged(user.Id, Offset(page, pageSize), pageSize);
        if (favs.Count == 0)
            return new Paginated<CapsuleListItem>(new List<CapsuleListItem>(), Page(total, page, pageSize));

        var capsuleMap = (await _capsules.FindAllByIds(favs.Select(f => f.CapsuleId).ToList()))
            .ToDictionary(c => c.Id);
        var owners = (await _users.FindAllByIds(capsuleMap.Values.Select(c => c.OwnerId).ToList()))
            .ToDictionary(u => u.Id);

        var items = new List<CapsuleListItem>();
        foreach (var f in favs)
        {
            if (!capsuleMap.TryGetValue(f.CapsuleId, out var c)) continue;
            if (!owners.TryGetValue(c.OwnerId, out var owner)) continue;
            items.Add(_mapper.ListItem(c, owner, favoritedByMe: true, favoritedAt: f.CreatedAt));
        }
        return new Paginated<CapsuleListItem>(items, Page(total, page, pageSize));
    }

    private async Task<HashSet<Guid>> LoadFavoriteSet(Guid? viewerId, IReadOnlyList<Guid> capsuleIds)
    {
        if (viewerId is null || capsuleIds.Count == 0) return new HashSet<Guid>();
        var favs = await _favorites.FindByUserAndCapsuleIds(viewerId.Value, capsuleIds);
        return favs.Select(f => f.CapsuleId).ToHashSet();
    }

    private static int Offset(int page, int pageSize) => (page - 1) * pageSize;

    private static Pagination Page(long total, int page, int pageSize)
    {
        var totalPages = pageSize == 0 ? 0 : (int)Math.Ceiling((double)total / pageSize);
        return new Pagination(page, pageSize, total, totalPages);
    }
}
